File: plane_parallel/plane.py
# =============================================================================
# Radiative Transfer of Photons: plane parallel atmosphere.
# Photons will be isotropically scattered, so every intereaction of a photon
# will be a scattering event.
# =============================================================================

import numpy as np
import timeit
from matplotlib import pyplot as plt


def sample_optical_depth(tau_max):
    """
    Randomly samples an optical depth and distance travelled for a photon.

    Parameters
    ----------
    tau_max: float.
        The maximum optical depth in the parallel plane.

    Returns
    -------
    tau: float.
        The sampled value of the optical depth.
    L: float.
        The distance the photon will travel due to the optical depth sampled.
    """

    xi = np.random.random()
    tau = -np.log(xi)
    L = tau/tau_max

    return tau, L


def emit_photon():
    """
    Generates a photon from the origin of a system (0, 0, 0) with a random
    direction.

    Returns
    -------
    x: float.
        The x coordinate of the photon.
    y: float.
        The y coordinate of the photon.
    z: float.
        The z coorindate of the photon.
    costheta: float.
        The cosine of the theta direction.
    sintheta: float.
        The sine of the theta direction.
    cosphi: float.
        The cosine of the phi direction.
    sinphi: float.
        The sine of the phi direction.
    """

    xi1 = np.random.random()
    xi2 = np.random.random()
    # generate a photon's angles using random sampling
    costheta = np.sqrt(xi1)
    sintheta = np.sqrt(1 - costheta ** 2)  # cos ** 2 + sin ** 2 = 1
    phi = 2 * np.pi * xi2
    cosphi = np.cos(phi)
    sinphi = np.sin(phi)
    # set the photon to be at the origin of a cartersian grid
    x, y, z = 0, 0, 0

    return x, y, z, costheta, sintheta, cosphi, sinphi


def isotropic_scattering():
    """
    Generates a new direction for a photon which has been isotropically
    scattered.

    Returns
    -------
    costheta: float.
        The cosine of the scattered theta direction.
    sintheta: float.
        The sine of the scattered theta direction.
    cosphi: float.
        The cosine of the scattered phi direction.
    sinphi: float.
        The sine of the scattered the phi direction.
    """

    xi1 = np.random.random()
    xi2 = np.random.random()

    costheta = 2 * xi1 - 1
    sintheta = np.sqrt(1 - costheta ** 2)
    phi = 2 * np.pi * xi2
    cosphi = np.cos(phi)
    sinphi = np.sin(phi)

    return costheta, sintheta, cosphi, sinphi, phi


def bin_photons(mu, mu_bins, n_photons):
    """
    Bins the photons into corresponding angle bins corresponding to the
    direction they left the slab.

    Parameters
    ----------
    mu: (1 x n_photons) array of floats.
        The angle mu = cos(theta) which the photons left the slab.
    mu_bins: integer.
        The number of angle bins to bin for. The angle bins are from
        0 < mu < 1, hence 0 < theta < 90
    n_photons: integer.
        The number of photons in the MC simulation.

    Returns
    -------
    mu_hist: (1 x mu_bins) array of ints.
        An array containing the number of photons in the binned angles, i.e.
        the number of photon packets which leave the plane in the binned angle
        range.
    theta: (1 x mu_bins) array of floats.
        An array containing the theta angles for which the photons have been
        binned.
    """

    dtheta = 1/mu_bins
    half_width = 0.5 * dtheta  # see kenny wood's code for half_width...?

    theta = np.zeros(mu_bins)
    mu_hist = np.zeros(mu_bins)

    # calculate the theta binning angles
    for i in range(mu_bins):
        theta[i] = np.arccos(i * dtheta + half_width) * (180/np.pi)

    # bin the mu = cos(theta) angles accordingly
    for i in range(n_photons):
        j = abs(int(mu[i] * mu_bins))
        mu_hist[j] = mu_hist[j] + 1

    return mu_hist, theta


def moments(mu_bins, mu_hist, theta):
    """
    Calculate the intensity of the radiation field by counting photons.

    Parameters
    ----------
    mu_bins: integer.
        The number of bins in the range 0 - 90 degrees.
    mu_hist: (1 x mu_bins) array of ints.
        An array containing the number of photons in the binned angles, i.e.
        the number of photon packets which leave the plane in the binned angle
    theta: (1 x mu_bins) array of floats.
        An array containing the theta angles for which the photons have been
        binned.

    Returns
    -------
    intensity: (1 x mu_bins) array of floats.
        The intensity estimated via the number of photon in each binned angle.
    """

    intensity = np.zeros(mu_bins)

    for i in range(mu_bins):
        intensity[i] = (mu_hist[i] * mu_bins)/(2 * np.sum(mu_hist) *
                                               np.cos(theta[i] * np.pi/180))

    return intensity


# =============================================================================
# Simulation Parameters
# =============================================================================

start = timeit.default_timer()

# set simulation parameters, 10e4 photon packets is ~5 minutes
n_photons = int(10e4)
tau_max = 17
mu_bins = 20
albedo = 1

MU = np.zeros(n_photons)
PHI = np.zeros(n_photons)

if n_photons:
    photon_count = 1
    while photon_count <= n_photons:
        # genereate a photon at the origin of the system
        xt, yt, zt, costheta, sintheta, cosphi, sinphi = emit_photon()

        # a photon will scatter at random angles and optical depths until it
        # is either absorbed (terminated) or exits the area
        while (zt >= 0) and (zt <= 1):
            # assign original positions for photons
            x = xt
            y = yt
            z = zt
            # sample a random optical depth and travel distance
            tau, L = sample_optical_depth(tau_max)
            # update the positions of the photon
            xt = x + L * sintheta * cosphi
            yt = y + L * sintheta * sinphi
            zt = z + L * costheta

            # determine if the photon scatters or not and update direction
            xi = np.random.rand()
            if (xi < albedo) and (zt > 0) and (zt < 1):
                costheta, sintheta, cosphi, sinphi, phi = \
                    isotropic_scattering()

        # if zt < 0, the photon has left through the bottom of the plane, and
        # gone into the star therefore restart the packet
        # if zt > 1, the photon has left through the top of the plane, record
        # the direction the photon left
        if (zt >= 1.0):
            i = photon_count - 1  # python uses zero indexing so etc
            MU[i] = costheta
            PHI[i] = phi

            photon_count += 1  # increment photon counter :^)

    mu_hist, theta = bin_photons(MU, mu_bins, n_photons)
    intensity = moments(mu_bins, mu_hist, theta)

    # plot the intensity against angle
    plt.figure(figsize=(8, 8))
    plt.plot(theta, intensity, 'o')
    plt.xlabel(r'Angle, $\theta$')
    plt.ylabel(r'Mean Intensity, $J$')
    plt.savefig('intensity.pdf')
    plt.show()

    stop = timeit.default_timer()
    print("Scattering completed in: {:3.2f} seconds.".format(stop - start))
else:
    stop = timeit.default_timer()
    print("Scattering  of zero photons completed in: {:3.2f} seconds."
          .format(stop - start))

File: plane.c
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <time.h>

#define N_PHOTONS 1000000
#define TAU_MAX 7
#define ALBEDO 1
#define MU_BINS 10
#define PI 3.14159265359

/*-------------------------- struct definitions -----------------------------*/

/* photon packet */
struct photon {
    double x;
    double y;
    double z;
    double costheta;
    double sintheta;
    double cosphi;
    double sinphi;
    double phi;
};

/* histogram for binned angles and count */
struct photon_hist {
    int bins[MU_BINS];
    double theta[MU_BINS];
};

/*---------------------------------------------------------------------------*/


/*-------------- function prototypes of all the functions -------------------*/

/* generates random numbers */
double random_number(void);
/* returns a photon struct which is a photon pointing in a random direction
   at the origin of the slab */
struct photon *emit_photon(void);
/* isotropically scatters a photon packet */
int *scatter_photon(struct photon *packetPtr);
/* bins the escape angles of the photon packets and creates a histogram */
struct photon_hist *bin_photons(double mu[]);
/* calculates the moments of the radiation field */
int *calculate_intensity(struct photon_hist *hist, double *p);
/* writes out the data to file */
int write_to_file(struct photon_hist *histPtr, double *intensity);

/*---------------------------------------------------------------------------*/

/* ---------------------------- main function -------------------------------*/
int main(void){

    /* initalise all of the variables */
    int photon_count, i;
    double MU[N_PHOTONS];
    double PHI[N_PHOTONS];
    double intensity[MU_BINS];
    struct photon packet;
    struct photon_hist hist;
    char* c_time_string;
    time_t current_time;

    /* initalise all of the pointers and point them at something */
    struct photon *packetPtr;
    packetPtr = &packet;
    struct photon_hist *histPtr;
    histPtr = &hist;
    double *intensityPtr;
    intensityPtr = intensity;

    current_time = time(NULL);
    c_time_string = ctime(&current_time);

    printf("Beginning simulation...\n");
    printf("Current time: %s\n", c_time_string);

    /* set the seed for rand() */
    srand(time(NULL));
    
    for (photon_count = 0; photon_count < N_PHOTONS; photon_count++){
        /* emit, scatter the photon packet and then record its escape angle */
        *packetPtr = *emit_photon();
        *scatter_photon(packetPtr);
        MU[photon_count] = packetPtr -> costheta;

        /* provides a progress report */
        if ((photon_count + 1) % 100000 == 0){
            printf("%6.0d photon packets transported (%3.0f%%)\n",
            (photon_count + 1),(photon_count + 1.0)/N_PHOTONS * 100.0);
        }
    }

    /* bin the photons and calculate the intensity */
    *histPtr = *bin_photons(MU);
    calculate_intensity(histPtr, intensityPtr);
    write_to_file(histPtr, intensity);

    printf("\nSimulation completed.\n");
    return 0;
}

/*---------------------------------------------------------------------------*/

/* return a random number between 0 and 1 */
double random_number(void){

    return (double)rand()/(double)RAND_MAX;
}

/* create a photon packet in a random direction at the origin */
struct photon *emit_photon(void){
    double xi1, xi2;
    double costheta, phi;
 
    struct photon *packetPtr;
    struct photon packet;
    packetPtr = &packet;
 
    /* generate 2 random numbers and the two directions */
    xi1 = random_number();
    xi2 = random_number();
    costheta = sqrt(xi1);
    phi = 2 * M_PI * xi2;
 
    /* populate the photon struct with values */
    packetPtr -> x = 0.0;
    packetPtr -> y = 0.0;
    packetPtr -> z = 0.0;
    packetPtr -> costheta = costheta;
    packetPtr -> sintheta = sqrt(1 - costheta * costheta);
    packetPtr -> cosphi = cos(phi);
    packetPtr -> sinphi = sin(phi);
    packetPtr -> phi = phi;
 
    return packetPtr;
}

/* isotropically scatter a photon */
int *scatter_photon(struct photon *packetPtr){
    double tau, L, xi;
 
    while (packetPtr -> z >= 0.0 && packetPtr -> z <= 1.0){
        /* sample a random optical depth and calculate the dist L */
        tau = -1.0 * log(random_number());
        L = tau/TAU_MAX;
 
        /* update the position of the packet */
        packetPtr -> x += L * packetPtr -> sintheta * packetPtr -> cosphi;
        packetPtr -> y += L * packetPtr -> sintheta * packetPtr -> sinphi;
        packetPtr -> z += L * packetPtr -> costheta;
 
        /* if z < 0, has gone into the atmosphere therefore restart the
           packet */
        if (packetPtr -> z < 0.0){
            *packetPtr = *emit_photon();
        }
 
        xi = random_number();
        /* scatter the photon is random number is less than albedo */
        if (xi < ALBEDO && packetPtr -> z < 1.0){
            packetPtr -> costheta = 2 * random_number() -1;
            packetPtr -> sintheta = sqrt(1 - (packetPtr -> costheta) *
                                        (packetPtr -> costheta));
            packetPtr -> phi = 2 * M_PI * random_number();
            packetPtr -> cosphi = cos(packetPtr -> phi);
            packetPtr -> sinphi = sin(packetPtr -> phi);
        }
    }
 
    return 0;
}

/* bin the packet escape angles */
struct photon_hist *bin_photons(double mu[]){
    int i, bin_index;
    double dtheta, half_width;
 
    struct photon_hist *histPtr;
    struct photon_hist hist;
    histPtr = &hist;
 
    dtheta = 1.0/MU_BINS;
    half_width = 0.5 * dtheta;
 
    /* calculate the binned angles */
    for (i = 0; i < MU_BINS; i++){
        histPtr -> bins[i] = 0; // set the bin count to zero
        histPtr -> theta[i] = acos(i * dtheta + half_width) * (180/M_PI);
    }
 
    /* bin the escape angles accordinly */
    for (i = 0; i < N_PHOTONS; i++){
        bin_index = abs((int)(mu[i] * MU_BINS));
        histPtr -> bins[bin_index] += 1;
    }
 
    return histPtr;
}

/* calculate the intensity */
int *calculate_intensity(struct photon_hist *hist, double *p){
    int i;
 
    for (i = 0; i < MU_BINS; i++){
        p[i] = ((float)hist -> bins[i] * MU_BINS)/(2 * N_PHOTONS * cos(
            hist -> theta[i] * M_PI/180));
    }
 
    return 0;
}

/* output the data to file */
int write_to_file(struct photon_hist *histPtr, double *intensity){
    int i;

    FILE *writeFile;

    /* write the output to file */
    if ((writeFile = fopen("intensity.dat", "w")) == NULL){
        printf("Cannot open 'intensity.dat' for writing.");
        return -1;
    }

    fprintf(writeFile, "%s", "theta\tcounts\tintensity\n");
    for (i = 0; i < MU_BINS; i++){
        fprintf(writeFile, "%f\t%d\t%f\n", histPtr -> theta[i],
                histPtr -> bins[i], intensity[i]);
    }

    if (fclose(writeFile) != 0){
        printf("File could not be closed.");
        return -1;
    }

    return 0;
}
